File: src/com/shape/ast/SyntaxTree.java
package com.shape.ast;

import com.shape.ast.nodes.Node;

public class SyntaxTree {
    private Node root;
    private Node current;

    public SyntaxTree() {
    }

    public void leave() {
        current = current.getParent();
    }

    public void addNode(Node node) {
        if (root == null) root = current = node;
        else current.addChild(node);
    }

    public void visit(Node node) {
        if (node.getParent() != null && node.getParent() != current)
            throw new IllegalArgumentException("you can only visit children of current node");
        current = node;
    }

    public Node getRoot() {
        return root;
    }

    public void visitNodes(RecursiveCallback callback) {
        callback.visit(root);
        visitNodes(root, callback);
    }

    private void visitNodes(Node node, RecursiveCallback callback) {
        for (Node child : node.getChildren()) {
            callback.visit(child);
            visitNodes(child, callback);
        }
    }

    public interface RecursiveCallback {
        void visit(Node node);
    }
}

File: compiler/src/main/java/com/shape/ast/SyntaxTree.kt
package com.shape.ast

import com.shape.ast.nodes.Node

/**
 * Copyright (C) 2015 Therne. All rights are reserved.
 * @author Vista
 */
class SyntaxTree {
    var current: Node? = null
        private set

    var root: Node? = null
        private set

    fun addNode(node: Node) {
        if (root == null) {
            root = node
            current = node
        }
        else current?.addChild(node)
    }

    fun visit(node: Node) {
        if (node.parent != null && node.parent !== current)
            throw IllegalArgumentException("you can only visit children of current node")
        current = node
    }

    fun visitNodes(callback: (Node?) -> Unit) {
        callback(root as Node)
        visitNodes(root as Node, callback)
    }

    fun leave() {
        current = current?.parent
    }

    private fun visitNodes(node: Node, callback: (Node?) -> Unit) {
        for (child in node.children) {
            callback(child)
            visitNodes(child, callback)
        }
    }
}